File: P1151 Minimum Swaps to Group All 1's Together/solution.py
class Solution:
    # Sliding Window with Two Pointers
    # Time complexity: O(n), when n is the length of the array.
    # Space complexity: O(1).
    def minSwaps(self, data: List[int]) -> int:
        ones_in_data = sum(data)
        ones_in_the_window = max_ones_in_the_window = 0
        left = right = 0
        
        while right < len(data):
            ones_in_the_window += data[right]
            right += 1
            if right - left > ones_in_data:
                ones_in_the_window -= data[left]
                left += 1
            max_ones_in_the_window = max(max_ones_in_the_window, ones_in_the_window)
        
        return ones_in_data - max_ones_in_the_window

File: P1151 Minimum Swaps to Group All 1's Together/Solution.java
class Solution {
    // Sliding Window with Two Pointers
    // Time complexity: O(n), when n is the length of the array.
    // Space complexity: O(1).
    public int minSwaps(int[] data) {
        int onesInData = 0;
        for (int i = 0; i < data.length; i++) {
            onesInData += data[i];
        }
        int onesInTheWindow = 0, maxOnesInTheWindow = 0;
        int left = 0, right = 0;
        
        while (right < data.length) {
            onesInTheWindow += data[right];
            right++;
            if (right - left > onesInData) {
                onesInTheWindow -= data[left];
                left++;
            }
            maxOnesInTheWindow = Math.max(maxOnesInTheWindow, onesInTheWindow);
        }
        return onesInData - maxOnesInTheWindow;
    }
}